Sources/SourceKitLSP/SourceKitLSPServer.swift

```diff
@@ -2128,27 +2128,44 @@ extension SourceKitLSPServer {
     callableUsrs += index.occurrences(ofUSR: data.usr, roles: .overrideOf).flatMap { occurrence in
       occurrence.relations.filter { $0.roles.contains(.overrideOf) }.map(\.symbol.usr)
     }
+    // callOccurrences are all the places that any of the USRs in callableUsrs is called.
+    // We also load the `calledBy` roles to get the method that contains the reference to this call.
     let callOccurrences = callableUsrs.flatMap { index.occurrences(ofUSR: $0, roles: .calledBy) }
-    let calls = callOccurrences.flatMap { occurrence -> [CallHierarchyIncomingCall] in
-      guard let location = indexToLSPLocation(occurrence.location) else {
-        return []
+
+    // Maps functions that call a USR in `callableUSRs` to all the called occurrences of `callableUSRs` within the
+    // function. If a function `foo` calls `bar` multiple times, `callersToCalls[foo]` will contain two call
+    // `SymbolOccurrence`s.
+    // This way, we can group multiple calls to `bar` within `foo` to a single item with multiple `fromRanges`.
+    var callersToCalls: [Symbol: [SymbolOccurrence]] = [:]
+
+    for call in callOccurrences {
+      // Callers are all `calledBy` relations of a call to a USR in `callableUsrs`, ie. all the functions that contain a
+      // call to a USR in callableUSRs. In practice, this should always be a single item.
+      let callers = call.relations.filter { $0.symbol.kind.isCallable }.map(\.symbol)
```

Comment:
I think it would be better to check the role here, with https://github.com/apple/swift/pull/72930 it ought to be sufficient to check for `.containedBy`, which would handle the case where the parent is a property (note that `isCallable` would be `false` in that case). Though for compatibility with older toolchains it may be better to check if the role contains `.calledBy`, and if that's empty, fallback to checking for `.containedBy`.

Reply:
Good suggestion with checking `$0.roles`.

There is a subtle difference between `calledBy` and `containedBy` relations.

```swift
func foo() {}

var testVar: Int {
  get {
    let myVar = foo()
    return 2
  }
}

func test() {
  _ = testVar
}
```

The call to `foo` in `let myVar = foo()` is called by `s:4test0A3VarSivg` (demangled `test.testVar.getter : Swift.Int`, display name `getter:testVar`) while it is contained by `s:4test0A3VarSivg02myB0L_ytvp` (demangled `myVar #1 : () in test.testVar.getter : Swift.Int` / display name `testVar`). But `s:4test0A3VarSivg02myB0L_ytvp` doesn’t have any callers while `s:4test0A3VarSivg` does show the call inside `test()`. So I’ll go with just checking the `calledBy` relation. Or do you see any advantage of `containedBy` over `calledBy`?

Reply:
> while it is contained by `s:4test0A3VarSivg02myB0L_ytvp`

That should have been fixed by https://github.com/apple/swift/pull/72930, is it still an issue? The main advantage of `containedBy` is that it works for cases like:

```swift
struct Foo {
  var member: Int = {
    return bar()
  }()

  func bar() {}
}
```

Even though `member` isn't a callable symbol, it probably makes sense to consider it the caller.

Reply:
Oh, right, sorry, you mentioned that. I was using an open source toolchain snapshot.

Let me try again once use have a toolchain snapshot with https://github.com/apple/swift/pull/72930 because I can’t be bothered to build a toolchain myself.

Reply:
That seems like a bug to me, the container there should be the same as the calledBy relation. The main advantage is that it handles one of the cases I sent above, ie.
```
func foo() -> Int {}

var testVar: Int = {
  foo()
}()
```

`foo` isn't really called by `testVar`, but IMO it's what a user would expect to see. Alternatively maybe we should have a `calledBy` to the init, but then that could also be very confusing. If we do the former I don't think we should continue the call hierarchy (though open to opinions there). The latter would make sense to continue though.

Reply:
:( I wish GitHub would tell me to refresh.

Reply:
Should we stop the traversal if we use the container instead? Does it make sense to continue from `member` there? Or should we have a calledBy to the `init`?